File: csharp/secret-handshake/SecretHandshake.cs
using System.Collections.Generic;
using System.Linq;

public static class SecretHandshake
{
    public static string[] Commands(int commandValue)
    {
        var commands = GenerateCommands(commandValue);
        return ((commandValue & 0b10000) != 0 ? commands.Reverse() : commands).ToArray();
    }

    private static IEnumerable<string> GenerateCommands(int commandValue)
    {
        for (int ii = 0; ii < 4; ii++)
        {
            var cmdVal = commandValue & (1 << ii);
            switch (cmdVal)
            {
                case 0b1:
                    yield return "wink";
                    break;
                case 0b10:
                    yield return "double blink";
                    break;
                case 0b100:
                    yield return "close your eyes";
                    break;
                case 0b1000:
                    yield return "jump";
                    break;
            }
        }
    }
}

File: haskell/secret-handshake/src/SecretHandshake.hs
module SecretHandshake (handshake) where

import Data.Bits

handshake :: Int -> [String]
handshake n = forwardOrBack n $ filter (/= "") $ map ($ n) secretSigns

secretSigns :: [(Int -> String)]
secretSigns = [0x1 `codeToGesture` "wink", 0x2 `codeToGesture` "double blink", 0x4 `codeToGesture` "close your eyes", 0x8 `codeToGesture` "jump"]

forwardOrBack :: Int -> ([a] -> [a])
forwardOrBack n
    | n .&. 0x10 == 0x10 = reverse
    | otherwise = id

codeToGesture :: Int -> String -> Int -> String
codeToGesture code gesture handshake'
    | code .&. handshake' == code = gesture
    | otherwise = ""